vyperlang/vyper
# fix[codegen]: fix overcopying of bytes in `make_setter`

### What I did

### How I did it

### How to verify it
check codesize improvements

note this PR originally had changes to the memmerge pass in venom, but they were superseded by the changes in https://github.com/vyperlang/vyper/pull/4422

### Commit message

```
in `_complex_make_setter`, when a dynamic type (e.g. `Bytes` or
`DynArray`) is contained within a tuple, the current code generation
copies the entire buffer without checking how large it is.

note this represents a gas issue since more bytes might be copied than
is necessary, but not a correctness issue.

in comparison, the `make_setter` implementations for `Bytes`
and `DynArray` limit the copy length to the runtime sizes of the
`Bytes`/`DynArray`.

this commit disables the full buffer copy and adds a heuristic to
the `make_setter` implementations for `DynArray` and `Bytes` so that
in certain cases, they copy the full buffer instead of checking the
length.
```

### Description for the changelog

### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->]()
